In [13]:
import numpy as np, pandas as pd
pd.set_option('display.max_column', None)

#import matplotlib.pyplot as plt, seaborn as sns
#size = 8
#plt.figure(figsize = (size, size))
#sns.set_style('whitegrid')

# API-Key Needed Packages

https://backtesting.tistory.com/entry/Available-Django-Authentication-Methods-for-Automated-Trading-Systems

https://www.quantconnect.com/

Some Packages need API keys to load data. You should use your own API key. 

* CCXT
* OpenDartReader
* Quandl -> Nasdaq Data Link
* Polygon.io
* Alpha Vantage

In [3]:
# CCXT
binance_api_key = ''
binance_secret_key = ''

# ODR
odr_api_key = ''


# polygon.io

# alpha vantage
av_api_key = '' # free

# NDL
NASDAQ_DATA_LINK_API_KEY = ''

# Nowdays Methods

## yfinance

crawl data from yahoo finance.

* https://github.com/ranaroussi/yfinance
* https://ranaroussi.github.io/yfinance/index.html

We can use this library mainly for US stock data.

In [2]:
#%pip install yfinance --upgrade --no-cache-dir

In [3]:
import yfinance as yf

msft = yf.Ticker("MSFT")

# get all stock info
msft.info

# get historical market data
hist = msft.history(period="1mo")
hist

Open        High         Low       Close  \
Date                                                                        
2024-11-21 00:00:00-05:00  419.500000  419.779999  410.290009  412.869995   
2024-11-22 00:00:00-05:00  411.369995  417.399994  411.059998  417.000000   
2024-11-25 00:00:00-05:00  418.380005  421.079987  414.850006  418.790009   
2024-11-26 00:00:00-05:00  419.589996  429.040009  418.850006  427.989990   
2024-11-27 00:00:00-05:00  425.109985  427.230011  422.019989  422.989990   
2024-11-29 00:00:00-05:00  420.089996  424.880005  417.799988  423.459991   
2024-12-02 00:00:00-05:00  421.570007  433.000000  421.309998  430.980011   
2024-12-03 00:00:00-05:00  429.839996  432.470001  427.739990  431.200012   
2024-12-04 00:00:00-05:00  433.029999  439.670013  432.630005  437.420013   
2024-12-05 00:00:00-05:00  437.920013  444.660004  436.170013  442.619995   
2024-12-06 00:00:00-05:00  442.299988  446.100006  441.769989  443.570007   
2024-12-09 00:00:00-05:00  442.600006  448.329987  440.500000  446.019989   
2024-12-10 00:00:00-05:00  444.390015  449.619995  441.600006  443.329987   
2024-12-11 00:00:00-05:00  444.049988  450.350006  444.049988  448.989990   
2024-12-12 00:00:00-05:00  449.109985  456.160004  449.109985  449.559998   
2024-12-13 00:00:00-05:00  448.440002  451.429993  445.579987  447.269989   
2024-12-16 00:00:00-05:00  447.269989  452.179993  445.279999  451.589996   
2024-12-17 00:00:00-05:00  451.010010  455.290009  449.570007  454.459991   
2024-12-18 00:00:00-05:00  451.320007  452.649994  437.019989  437.390015   
2024-12-19 00:00:00-05:00  441.619995  443.179993  436.320007  437.029999   
2024-12-20 00:00:00-05:00  433.109985  443.739990  428.630005  436.600006   

                             Volume  Dividends  Stock Splits  
Date                                                          
2024-11-21 00:00:00-05:00  20780200       0.83           0.0  
2024-11-22 00:00:00-05:00  24814600       0.00           0.0  
2024-11-25 00:00:00-05:00  27691100       0.00           0.0  
2024-11-26 00:00:00-05:00  23458900       0.00           0.0  
2024-11-27 00:00:00-05:00  18332400       0.00           0.0  
2024-11-29 00:00:00-05:00  16271900       0.00           0.0  
2024-12-02 00:00:00-05:00  20207200       0.00           0.0  
2024-12-03 00:00:00-05:00  18302000       0.00           0.0  
2024-12-04 00:00:00-05:00  26009400       0.00           0.0  
2024-12-05 00:00:00-05:00  21697800       0.00           0.0  
2024-12-06 00:00:00-05:00  18821000       0.00           0.0  
2024-12-09 00:00:00-05:00  19144400       0.00           0.0  
2024-12-10 00:00:00-05:00  18469500       0.00           0.0  
2024-12-11 00:00:00-05:00  19200200       0.00           0.0  
2024-12-12 00:00:00-05:00  20834800       0.00           0.0  
2024-12-13 00:00:00-05:00  20177800       0.00           0.0  
2024-12-16 00:00:00-05:00  23598800       0.00           0.0  
2024-12-17 00:00:00-05:00  22733500       0.00           0.0  
2024-12-18 00:00:00-05:00  24444500       0.00           0.0  
2024-12-19 00:00:00-05:00  22963700       0.00           0.0  
2024-12-20 00:00:00-05:00  64235200       0.00           0.0

## CCXT

CCXT : CryptoCurrency eXchange Trading Library

JAVAScript/Python 지원. 여러 암호화폐 거래소와 통합되는 라이브러리 및 실시간 데이터 처리 가능.

국내 거래소 중 upbit, bithumb을 지원하며, 국외 거래소에서는 binance 등을 지원.

https://github.com/ccxt/ccxt

In [8]:
#%pip install ccxt --upgrade --no-cache-dir

In [ ]:
import ccxt
print(ccxt.exchanges) # print a list of all available exchange classes

['ace', 'alpaca', 'ascendex', 'bequant', 'bigone', 'binance', 'binancecoinm', 'binanceus', 'binanceusdm', 'bingx', 'bit2c', 'bitbank', 'bitbay', 'bitbns', 'bitcoincom', 'bitfinex', 'bitfinex2', 'bitflyer', 'bitget', 'bithumb', 'bitmart', 'bitmex', 'bitopro', 'bitpanda', 'bitrue', 'bitso', 'bitstamp', 'bitteam', 'bitvavo', 'bl3p', 'blockchaincom', 'blofin', 'btcalpha', 'btcbox', 'btcmarkets', 'btcturk', 'bybit', 'cex', 'coinbase', 'coinbaseadvanced', 'coinbaseexchange', 'coinbaseinternational', 'coincheck', 'coinex', 'coinlist', 'coinmate', 'coinmetro', 'coinone', 'coinsph', 'coinspot', 'cryptocom', 'currencycom', 'delta', 'deribit', 'digifinex', 'exmo', 'fmfwio', 'gate', 'gateio', 'gemini', 'hitbtc', 'hitbtc3', 'hollaex', 'htx', 'huobi', 'huobijp', 'hyperliquid', 'idex', 'independentreserve', 'indodax', 'kraken', 'krakenfutures', 'kucoin', 'kucoinfutures', 'kuna', 'latoken', 'lbank', 'luno', 'lykke', 'mercado', 'mexc', 'ndax', 'novadax', 'oceanex', 'okcoin', 'okx', 'onetrading', 'oxfun

In [ ]:
import ccxt

# 거래소 인스턴스 생성 (예: Binance)
exchange = ccxt.binance({
    'apiKey': binance_api_key,
    'secret': binance_secret_key,
})

# BTC/USDT 마켓의 현재 가격 조회
btc_usdt = exchange.fetch_ticker('BTC/USDT')
print(btc_usdt['last'])  # 마지막 거래 가격 출력

# 사용자 계좌 잔액 조회
balance = exchange.fetch_balance()
print(balance['total'])  # 총 잔액 출력

## pykrx

한국 주식 데이터 제공. 웹 스크래핑으로 가져오기에 시간 걸릴 수 있음.

https://github.com/sharebook-kr/pykrx

In [5]:
#%pip install pykrx --upgrade --no-cache-dir

In [6]:
import pykrx
import pykrx.stock as korstock
import pykrx.bond as korbond

tickers = korstock.get_market_ticker_list("20190225")
print(tickers)

['095570', '068400', '006840', '027410', '282330', '138930', '001460', '001465', '001040', '079160', '000120', '011150', '011155', '001045', '097950', '097955', '037560', '000590', '012030', '016610', '005830', '000990', '000995', '139130', '004840', '155660', '069730', '017940', '007700', '114090', '078930', '006360', '001250', '007070', '078935', '012630', '039570', '089470', '294870', '082740', '003560', '175330', '234080', '001060', '001067', '001065', '096760', '105560', '002380', '009070', '009440', '119650', '092220', '001390', '001940', '025000', '092230', '000040', '044450', '030200', '033780', '030210', '058850', '093050', '003550', '034220', '001120', '051900', '051905', '003555', '032640', '011070', '066570', '066575', '108670', '108675', '051910', '051915', '079550', '006260', '000680', '010120', '229640', '023150', '035420', '181710', '005940', '005945', '034310', '008260', '004250', '004255', '010060', '005490', '100840', '064960', '003570', '036530', '010950', '010955',

## marcap

국내 종목에 대한 시가총액 데이터셋 제공. 오프라인 다운로드 가능. 

https://github.com/FinanceData/marcap

아래 링크를 통해 매일 업데이트되는 marcap 데이터 다운로드 가능.

In [10]:
#!git clone "https://github.com/FinanceData/marcap.git" marcap

In [11]:
from marcap import marcap_data

# 특정 날짜 전종목
df = marcap_data('2021-01-21') 

# 특정 기간 전종목
df = marcap_data('2020-01-01', '2020-12-31') 

# 특정 기간 단일 종목
df = marcap_data('2021-01-01', '2021-01-31', code='005930') 

In [12]:
from marcap import marcap_data

# 특정 날짜를 지정하여 읽기
df = marcap_data('2021-01-21')

# 기간을 지정하여 읽기 (2018년 1년간 전종목)
df = marcap_data('2020-01-01', '2020-12-31')

# 기간을 지정하여, 특정 종목 읽기 (2018년 1년간 특정 종목)
df = marcap_data('2021-01-01', '2021-01-31', code='005930')

# 1995-05-02 ~ 2021-01-15일까지 25년 데이터를 모두
df = marcap_data('1995-05-02', '2021-01-15')

In [22]:
# ETF : cannot downloaded from here
df = marcap_data('2024-12-19', code='473460')
df

Empty DataFrame
Columns: [Code, Name, Close, Dept, ChangeCode, Changes, ChagesRatio, Volume, Amount, Open, High, Low, Marcap, Stocks, Market, MarketId, Rank]
Index: []

## FinanceDataReader

데이터 로딩 라이브러리. `pandas-datareader`를 보완하기 위한 목적으로 만들어짐.

https://github.com/FinanceData/FinanceDataReader?tab=readme-ov-file

https://financedata.github.io/posts/finance-data-reader-users-guide.html

In [31]:
#%pip install finance-datareader --upgrade --no-cache-dir

In [ ]:
import FinanceDataReader as fdr

# KOSPI Index 코스피 지수 데이터 
df = fdr.DataReader('IXIC', '2020') # 2020-01-01 ~ 현재
df

Open          High  ...     Adj Close      Volume
Date                                    ...                          
2020-01-02   9039.459961   9093.429688  ...   9092.190430  2862700000
2020-01-03   8976.429688   9065.759766  ...   9020.769531  2586520000
2020-01-06   8943.500000   9072.410156  ...   9071.469727  2810450000
2020-01-07   9076.639648   9091.929688  ...   9068.580078  2381740000
2020-01-08   9068.030273   9168.889648  ...   9129.240234  2472620000
...                  ...           ...  ...           ...         ...
2024-08-05  15712.530273  16453.460938  ...  16200.080078  6852190000
2024-08-06  16261.360352  16620.310547  ...  16366.849609  5908930000
2024-08-07  16622.310547  16709.810547  ...  16195.809570  5913620000
2024-08-08  16408.269531  16694.250000  ...  16660.019531  5815880000
2024-08-09  16636.519531  16789.220703  ...  16745.300781  5783410000

[1159 rows x 6 columns]

## OpenDartReader

금융감독원 전자공시 시스템 DART 로딩 라이브러리

* openDart 사이트 : https://opendart.fss.or.kr/
* OpenDartReader 공식 Github : https://github.com/FinanceData/OpenDartReader

In [29]:
#%pip install opendartreader --upgrade --no-cache-dir

In [ ]:
import OpenDartReader as odr
dart = odr(odr_api_key)

dart.list('000660')

corp_code corp_name stock_code  ...  flr_nm  rcept_dt rm
0     00164779    SK하이닉스     000660  ...     박진규  20240805   
1     00164779    SK하이닉스     000660  ...  국민연금공단  20240805   
2     00164779    SK하이닉스     000660  ...  SK하이닉스  20240731  공
3     00164779    SK하이닉스     000660  ...  SK하이닉스  20240726  유
4     00164779    SK하이닉스     000660  ...  SK하이닉스  20240726  유
...        ...       ...        ...  ...     ...       ... ..
2542  00164779    SK하이닉스     000660  ...  SK하이닉스  20000404   
2543  00164779    SK하이닉스     000660  ...  삼일회계법인  20000403   
2544  00164779    SK하이닉스     000660  ...  SK하이닉스  20000401   
2545  00164779    SK하이닉스     000660  ...  SK하이닉스  19990813   
2546  00164779    SK하이닉스     000660  ...  SK하이닉스  19990331  연

[2547 rows x 9 columns]

## pyfred

https://github.com/artdavis/pyfred

# Charged Data Channels

## 한국거래소

https://www.krx.co.kr/main/main.jsp

## Fama/French Data Library

https://mba.tuck.dartmouth.edu/pages/faculty/ken.french/data_library.html

## Factset

https://www.factset.com/

과거 금융시장 데이터에 대해 가장 신뢰도 높은 사이트 중 하나로 여겨짐. 

ClariFI를 인수하여 분석 서비스 역시 제공 중.

## Yardeni Research

https://yardeni.com/

## Bloomberg Terminal

블대리... 아시죠?

## Polygon.io

Polygon.io

https://polygon.io/docs/stocks/getting-started

Polygon.io의 무료 플랜을 통해 다음과 같은 데이터를 이용할 수 있습니다:

* 미국 주식 (U.S. Equities): 일일 종가 데이터와 최대 2년간의 분 단위 과거 데이터를 제공합니다.
* 외환 (Forex): 일일 종가 데이터와 최대 2년간의 분 단위 과거 데이터를 제공합니다.
* 암호화폐 (Cryptocurrency): 일일 종가 데이터와 최대 2년간의 분 단위 과거 데이터를 제공합니다.

이러한 데이터는 RESTful API를 통해 액세스할 수 있으며, 무료 플랜에서는 분당 최대 5회의 API 요청이 가능합니다. 

## AlphaVentage

https://www.alphavantage.co/documentation/

https://todo-wr-sth.tistory.com/5

https://algotrading101.com/learn/alpha-vantage-guide/?utm_source=chatgpt.com

https://www.quantconnect.com/docs/v2/local-platform/datasets/alpha-vantage

## tiingo

https://www.tiingo.com/

## Finazon

https://finazon.io/blog/finazon-rest-api-introduction?utm_source=chatgpt.com

## Quandl -> Nasdaq Data Link

NASDAQ data link와 연결.

* NASDAQ data link : https://data.nasdaq.com/
* github : https://github.com/Nasdaq/data-link-python

In [ ]:
#%pip install nasdaq-data-link

In [ ]:
import nasdaqdatalink as ndqlink
data = ndqlink.get('NSE/OIL')
df

DataLinkError: (Status 403) Something went wrong. Please try again. If you continue to have problems, please contact us at connect@data.nasdaq.com.

# Obsolete Methods

## pandas_datareader

pandas_datareader를 사용해서 data를 load하는 패키지. 현재는 사실상 유지/보수가 중단됨.

https://github.com/pydata/pandas-datareader/issues/977

* https://github.com/pydata/pandas-datareader
* https://pandas-datareader.readthedocs.io/en/latest/readers/yahoo.html

In [ ]:
#%pip install pandas-datareader

In [49]:
import pandas_datareader as pdr

# df = pdr.get_data_yahoo('AAPL')

야후 파이낸스로부터 데이터를 로드하는 방법은 사용 불가능.

In [46]:
from pandas_datareader.famafrench import get_available_datasets
import pandas_datareader.data as web

In [48]:
ds = web.DataReader('5_Industry_Portfolios', 'famafrench')
print(ds['DESCR'])

5 Industry Portfolios
---------------------

This file was created by CMPT_IND_RETS using the 202410 CRSP database. It contains value- and equal-weighted returns for 5 industry portfolios. The portfolios are constructed at the end of June. The annual returns are from January to December. Missing data are indicated by -99.99 or -999. Copyright 2024 Eugene F. Fama and Kenneth R. French

  0 : Average Value Weighted Returns -- Monthly (59 rows x 5 cols)
  1 : Average Equal Weighted Returns -- Monthly (59 rows x 5 cols)
  2 : Average Value Weighted Returns -- Annual (5 rows x 5 cols)
  3 : Average Equal Weighted Returns -- Annual (5 rows x 5 cols)
  4 : Number of Firms in Portfolios (59 rows x 5 cols)
  5 : Average Firm Size (59 rows x 5 cols)
  6 : Sum of BE / Sum of ME (6 rows x 5 cols)
  7 : Value-Weighted Average of BE/ME (6 rows x 5 cols)


C:\Users\Na\AppData\Local\Temp\ipykernel_24284\3709057983.py:1: FutureWarning: The argument 'date_parser' is deprecated and will be removed in a future version. Please use 'date_format' instead, or read your data in as 'object' dtype and then call 'to_datetime'.
  ds = web.DataReader('5_Industry_Portfolios', 'famafrench')
C:\Users\Na\AppData\Local\Temp\ipykernel_24284\3709057983.py:1: FutureWarning: The argument 'date_parser' is deprecated and will be removed in a future version. Please use 'date_format' instead, or read your data in as 'object' dtype and then call 'to_datetime'.
  ds = web.DataReader('5_Industry_Portfolios', 'famafrench')
C:\Users\Na\AppData\Local\Temp\ipykernel_24284\3709057983.py:1: FutureWarning: The argument 'date_parser' is deprecated and will be removed in a future version. Please use 'date_format' instead, or read your data in as 'object' dtype and then call 'to_datetime'.
  ds = web.DataReader('5_Industry_Portfolios', 'famafrench')
C:\Users\Na\AppData\Local\Te

단, 타 함수들은 일부 사용 가능한 측면 존재. 야후 파이낸스 사용시 `yfinance`를 대신 사용할 것.

## investpy/investiny

investing.com에서 크롤링하는 패키지.

* investing.com : https://www.investing.com/
* 패키지 공식 github 
    * `investpy` : https://investpy.readthedocs.io/_info/introduction.html
    * `investiny` : https://github.com/alvarobartt/investiny

* `investpy`는 investing.com에서 크롤링하는 패키지로, 현재 클라우드플레어를 통해 API를 가져올 수 없게 되어 업데이트 중단. 403 error 출력. 
* `investiny`는 클라우드플레어를 우회하여 API 제공이 가능하며, 더 빠르고 경량화된 버전. 마찬가지로 업데이트 중단.

In [ ]:
#%pip install investpy
#%pip install investiny

In [38]:
#import investpy

#df = investpy.get_stock_historical_data(stock='AAPL', country='United States', from_date='01/01/2010', to_date='01/01/2020')
#print(df.head())

In [39]:
#from investiny import historical_data

#data = historical_data(investing_id=6408, from_date="09/01/2022", to_date="10/01/2022") 
# Returns AAPL historical data as JSON (without date)